onnxruntime/core/optimizer/computation_reduction.cc

```diff
@@ -0,0 +1,271 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/safeint.h"
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/utils.h"
+#include "core/optimizer/computation_reduction.h"
+
+using namespace ONNX_NAMESPACE;
+using namespace ::onnxruntime::common;
+namespace onnxruntime {
+typedef std::function<Status(Graph&, Node&, Node&)> Handler;
+
+static const int GATHERND_BATCH_DIM = 1;
+
+static bool IsLeadingDimsEqual(const TensorShapeProto* input_shape, const TensorShapeProto* output_shape,
+                               int num_dim_to_check) {
+  ORT_ENFORCE(output_shape->dim_size() >= num_dim_to_check && input_shape->dim_size() >= num_dim_to_check);
+
+  for (int i = 0; i < num_dim_to_check; i++) {
+    auto& output_dim = output_shape->dim(i);
+    auto& input_dim = input_shape->dim(i);
+    if (output_dim.has_dim_value() && input_dim.has_dim_value()) {
+      if (output_dim.dim_value() != input_dim.dim_value()) {
+        return false;
+      }
+    } else if (output_dim.has_dim_param() && input_dim.has_dim_param()) {
+      if (output_dim.dim_param() != input_dim.dim_param()) {
+        return false;
+      }
+    } else {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+static int GetValidInputForGatherND(const Node& target_node) {
+  // If target_node's some input tensor have exactly same shape with
+  // target_node output tensor shape, then it is safe to gather using
+  // original slice ranges.
+  int candidate_input_index = -1;
+  auto output_shape = target_node.OutputDefs()[0]->Shape();
+  const int output_rank = output_shape->dim_size();
+  for (size_t i = 0; i < target_node.InputDefs().size(); i++) {
+    auto input_shape = target_node.InputDefs()[i]->Shape();
+    const int input_rank = input_shape->dim_size();
+    if (input_rank != output_rank) {
+      continue;
+    }
+
+    if (IsLeadingDimsEqual(input_shape, output_shape, GATHERND_BATCH_DIM + 1)) {
+      candidate_input_index = SafeInt<int>(i);
+      break;
+    }
+  }
+
+  return candidate_input_index;
+}
+
+static TensorShapeProto ReplaceSymbolicDimValue(const TensorShapeProto* shape, int replacement_axis,
+                                                std::string replacement_dim_value) {
```

Comment:
>std::string [](start = 48, length = 11)

Nit: let's use const std::string& to avoid re-construct a new string obj.

Reply:
Thanks. :)